---
title: "Monkey Pox Genetics May 2022"
output: html_notebook
---
# Introduction
[YouTube tutorial](https://www.youtube.com/watch?v=qhLSfx_wpeA) uploaded by [Dr. Danny Arends](https://dannyarends.nl/?).

Notebook completed by Samantha Martell

Opened: 29 May 2022

Final edit: 

## To-do list
1. Where do we get our data from? (e.g. USA, Portugal, Belgium). Source from NCBI. 
2. Load genome sequences from fasta and NCBI.
3. Create a blast database.
4. Load associated proteins from NCBI.
5. Create a protein model (Levenstein, MSA) -> consensus.
6. tblastn Consensus to viral genomes database.
7. Genomic layout of the virus.

### What is a consensus sequence?
A consensus sequence is a sequence of DNA, RNA, or protein that represents aligned, related sequences. The consensus sequence of the related sequences can be defined in different ways, but is normally defined by the most common nucleotide(s) or amino acid residue(s) at each position ([source](https://en.wikipedia.org/wiki/Consensus_sequence)).

# Initialising the notebook {.tabset}
## Set working directory
```{r}
# set working directory
setwd("C:/Users/samma/Monkeypox-genetics")
cat(getwd())
```
## Install packages
* Install packages as required. 
  * [Multiple Sequence Alignment](https://bioconductor.org/packages/release/bioc/html/msa.html)
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("msa")
```


## Importing packages
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("msa")
library(msa)
library(RColorBrewer)
library(rentrez)
library(stringdist)
library(seqinr)
```

# 0 Getting data from NCBI
* Go to [NCBI nucleotide database](https://www.ncbi.nlm.nih.gov/nuccore)
```{r}


