---
title: "Monkey Pox Genetics May 2022"
output: html_notebook
---
# Introduction
[YouTube tutorial](https://www.youtube.com/watch?v=qhLSfx_wpeA) uploaded by [Dr. Danny Arends](https://dannyarends.nl/?).

Notebook completed by Samantha Martell

Opened: 29 May 2022

Final edit: 

## To-do list
1. Where do we get our data from? (e.g. USA, Portugal, Belgium)
2. Load genome sequences from fasta and NCBI
3. Create a blast database
4. Load associated proteins from NCBI
5. Create a protein model (Levenstein, MSA) -> consensus
6. tblastn Consensus to viral genomes database
7. Genomic layout of the virus

# Initialising the notebook {.tabset}
## Set working directory
```{r}
# set working directory
setwd("C:/Users/samma/Monkeypox-genetics")
cat(getwd())
```
## Install packages
* Install packages as required. 
  * [Multiple Sequence Alignment](https://bioconductor.org/packages/release/bioc/html/msa.html)
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("msa")
```


## Importing packages
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("msa")
library(msa)
library(RColorBrewer)
library(rentrez)
library(stringdist)
library(seqinr)
```

# Testing
```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
