---
title: "Monkey Pox Genetics May 2022"
output: html_notebook
---
# Introduction

Notebook completed by Samantha Martell
First modified: 29 Apr 2022
Last modified: 04 Jun 2022
[YouTube tutorial](https://www.youtube.com/watch?v=qhLSfx_wpeA) uploaded by [Dr. Danny Arends](https://dannyarends.nl/?).
copyright (c) 2022 - Samantha Martell


## To-do list
1. Where do we get our data from? (e.g. USA, Portugal, Belgium). Source from NCBI. 
2. Load genome sequences from fasta and NCBI.
3. Create a blast database.
4. Load associated proteins from NCBI.
5. Create a protein model (Levenstein, MSA) -> consensus.
6. tblastn Consensus to viral genomes database.
7. Genomic layout of the virus.

### What is a consensus sequence?
A consensus sequence is a sequence of DNA, RNA, or protein that represents aligned, related sequences. The consensus sequence of the related sequences can be defined in different ways, but is normally defined by the most common nucleotide(s) or amino acid residue(s) at each position ([source](https://en.wikipedia.org/wiki/Consensus_sequence)).

# Initialising the notebook {.tabset}
## Clear environment
```{r}
rm(list = ls())
```

## Set working directory
```{r}
# set working directory
setwd("C:/Users/samma/Monkeypox-genetics/")
cat(getwd())
list.files()
```
## Install packages
* Install packages as required. 
  * [Multiple Sequence Alignment](https://bioconductor.org/packages/release/bioc/html/msa.html)
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("msa")
```


## Importing packages
```{r}
library(msa)
library(RColorBrewer)
library(rentrez)
library(stringdist)
library(seqinr)
library(rBLAST)
```

# 0. Getting data from NCBI
* Go to [NCBI nucleotide database](https://www.ncbi.nlm.nih.gov/nuccore)
* Search for monkeypox
* GenBank ID is ON563414.2
  * Relatively long at 197124 bp
* Should compare an (older) annotated genome with one of these newer genomes. 
* Danny has kindly put a .txt file containing information on various monkeypoxes and related viruses.
  * Based on set from [virological.org](https://virological.org/).
  
```{r}
sourceIDs <- read.csv(
  file = "data/monkeypox_annotation.txt", 
  sep = "\t", 
  header = TRUE, 
  row.names = 1
)

# check the dataframe
head(sourceIDs)
```
Check the structure of the new dataframe. 
```{r}
str(sourceIDs)
```

  
# 1. Load genome sequences
Use the function defined in `R/downloadData.R` to download data from the NCBI nucleotide database according to the sourceIDs given in Part 0.

```{r}
source("R/downloadData.R")
ncbi.seq <- downloadData(rownames(sourceIDs))
```

```{r}
class(ncbi.seq)
length(ncbi.seq)
str(ncbi.seq)
```

# 2. Creating a blast database
Use the function defined in `R/makeBlastDB.R` to write data downloaded in Part 1 to a file to avoid continuously accessing data from NCBI. 

```{r}
source("R/makeBlastDB.R")
db <- makeBlastDB(ncbi.seq)
```

## Writing to a file
Alternatively, write to .csv file since I can't get the. fsa file to read back in nicely. 
```{r}
write.csv(ncbi.seq, "data/DNAdb.csv")
```

Test the new file.
```{r}
source("R/readBlastDBasCsv.R")
ncbi.seq <- readBlastDBasCsv("data/DNAdb.csv")
```

```{r}
str(ncbi.seq)
```

# 3. Load associated proteins from NCBI
This code finds the proteins in the protein database which are annotated to the [NC_003310.1](https://www.ncbi.nlm.nih.gov/nuccore/NC_003310.1/) sequence. 

```{r}
source("R/queryProteins.R")
my.proteins <- queryProteins()
```

```{r}
str(my.proteins)
length(my.proteins)
```

We will query a few of these proteins - download a fraction of 191.  
```{r}
proteins <- c()
for (name in row.names(sourceIDs)){
  proteins <- c(proteins, queryProteins(name))
  Sys.sleep(5)
}
```

```{r}
length(proteins)
```

Write to a .csv file for later use.
```{r}
write.csv(proteins, "data/proteins.csv")
```
```{r}
source("R/readBlastDBasCsv.R")
proteins <- readBlastDBasCsv("data/proteins.csv")
str(proteins)
```

# 4. Create a protein model
## Exploring the data
Get the names of the proteins to have a look at them. 
```{r}
names.proteins <- unlist(names(proteins))
names.proteins[1:10]
```

Use `grep` to find a specific protein. 
```{r}
grep("Monoglyceride", names.proteins)
length(grep("Monoglyceride", names.proteins))
```

# Levenshtein  distance
Use the [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) to see how closely proteins are related.

For example, see which proteins have a Levenshtein distance of < 25 from the first protein. 
```{r}
which(stringdist(proteins[1], proteins, method = "lv") < 25)
```

Use this code to generate a subset of proteins similar to the first protein. 
```{r}
my.consensus.sequences <- c()
for(x in 1:5) {
  similar <- names(proteins[which(stringdist(proteins[x], proteins, method = "lv") < 15)])
  my.msa <- msa(AAStringSet(proteins[similar]))   # Build a multiple sequence alignment
  my.consensus <- msaConsensusSequence(my.msa)    # store the consensus sequence
  my.consensus.sequences <- c(my.consensus.sequences, my.consensus)
}
names(my.consensus.sequences) <- paste0("consensus_", 1:5)
```

Can use the consensus sequence to probe the monkeypox genome. 

# 5. `tblastn` consensus sequence to the viral genomes DB
```{r}
source("R/blast.R")
blast.res <- blast(my.consensus.sequences)
positions <- c(as.numeric(blast.res[, "sstart"]), as.numeric(blast.res[, "send"]))

```

```{r}
str(blast.res)
class(blast.res)
unique(blast.res["qseqid"])
```
The below tells us some genes appear twice in the genome:
```{r}
head(blast.res)
```

# 6. Genomic layout of the virus
Visualise the data.

```{r}
positions <- c(as.numeric(blast.res[,"sstart"]), as.numeric(blast.res[,"send"]))
min(positions); max(positions)
```

Studying a single protein.
  * Some strains only have a single copy of the gene. 
  * Some strains are missing the gene entirely.
  

```{r}
jpeg(file = "images/saving_plot1.jpg")

colz <- brewer.pal(5, "Set2")

op <- par(mar = c(10, 5, 5, 5))

plot(
  c(0, nrow(sourceIDs)), 
  y = c(min(positions), max(positions)), 
  t = 'n', 
  xaxt = 'n', 
  xlab = "",
  ylab = ""
)

for(name in rownames(sourceIDs)){
  genome <- which(blast.res[, "sseqid"] == name)
  for(gene in genome){
    y.start <- blast.res[gene, "sstart"]
    y.end <- blast.res[gene, "send"]
    xpos <- which(rownames(sourceIDs) == name)
    
    q <- blast.res[gene, "qseqid"]
    n <- unlist(lapply(strsplit(q, "_"), "[", 2))

    points(c(xpos, xpos), c(y.start, y.end), t = 'p', pch = 19, col = colz[as.numeric(n)])
    points(c(xpos, xpos), c(y.start, y.end), t = 'l', col = colz[as.numeric(n)])
  }
}

display.names <- paste0(sourceIDs[, "Country"], " ", sourceIDs[, "Year"], " ", sourceIDs[, "Sub.Type"])
axis(1, at  = 1:nrow(sourceIDs), display.names, las = 2, cex.axis = 0.7)

dev.off()
```





