---
title: "Monkey Pox Genetics May 2022"
output: html_notebook
---
# Introduction
[YouTube tutorial](https://www.youtube.com/watch?v=qhLSfx_wpeA) uploaded by [Dr. Danny Arends](https://dannyarends.nl/?).

Notebook completed by Samantha Martell

Opened: 29 May 2022

Final edit: 

## To-do list
1. Where do we get our data from? (e.g. USA, Portugal, Belgium). Source from NCBI. 
2. Load genome sequences from fasta and NCBI.
3. Create a blast database.
4. Load associated proteins from NCBI.
5. Create a protein model (Levenstein, MSA) -> consensus.
6. tblastn Consensus to viral genomes database.
7. Genomic layout of the virus.

### What is a consensus sequence?
A consensus sequence is a sequence of DNA, RNA, or protein that represents aligned, related sequences. The consensus sequence of the related sequences can be defined in different ways, but is normally defined by the most common nucleotide(s) or amino acid residue(s) at each position ([source](https://en.wikipedia.org/wiki/Consensus_sequence)).

# Initialising the notebook {.tabset}
## Clear environment
```{r}
rm(list = ls())
```

## Set working directory
```{r}
# set working directory
setwd("C:/Users/samma/Monkeypox-genetics/")
cat(getwd())
list.files()
```
## Install packages
* Install packages as required. 
  * [Multiple Sequence Alignment](https://bioconductor.org/packages/release/bioc/html/msa.html)
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("msa")
```


## Importing packages
```{r}
library(msa)
library(RColorBrewer)
library(rentrez)
library(stringdist)
library(seqinr)
```

# 0 Getting data from NCBI
* Go to [NCBI nucleotide database](https://www.ncbi.nlm.nih.gov/nuccore)
* Search for monkeypox
* GenBank ID is ON563414.2
  * Relatively long at 197124 bp
* Should compare an (older) annotated genome with one of these newer genomes. 
* Danny has kindly put a .txt file containing information on various monkeypoxes and related viruses.
  * Based on set from [virology.org](https://www.pascv.org/default.aspx).
  
```{r}
sourceIDs <- read.csv(
  file = "C:/Users/samma/Monkeypox-genetics/data/monkeypox_annotation.txt", 
  sep = "\t", 
  header = TRUE, 
  row.names = 1
)

# check the dataframe
head(sourceIDs)
```
Check the structure of the new dataframe. 
```{r}
str(sourceIDs)
```

  
# 1 Load genome sequences
```{r}
rseq <- entrez_fetch("nucleotide", rownames(sourceIDs)[1], rettype = "fasta")
```

Check the data.
```{r}
cat(class(rseq), "\n\n")
cat(str(rseq), "\n\n")
```

## Cleaning the data
Make the output readable using `strsplit()` and `unlist()`:
```{r}
cat(class(strsplit(rseq, "\n")), "\n")
cat(class(unlist(strsplit(rseq, "\n"))))
```

Use [`grep`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/grep) function to find the header row.
```{r}
rseq <- entrez_fetch("nucleotide", rownames(sourceIDs)[1:3], rettype = "fasta")
rseq <- unlist(strsplit(rseq, "\n"))
header.lines <- grep(">", rseq)
header.lines
# This tells us the first sequence starts on row 2, 
# the second sequence starts on row 2821, 
# the third sequence starts on row 5639.
```


Make a new dataframe containing all the relevant positions:
```{r}
seq.pos <- cbind(
  start = header.lines + 1, # calls all the start positions
  stop = c(header.lines[-1] - 1, length(rseq)) # calls all the end positions + the final entry in header.lines
)
seq.pos
```


```{r}
# apply to every row in seq.pos function (x) (to be defined...) and store in seq.fasta
seq.fasta <- apply(seq.pos, 1, function(x){
  # debug cat statement
  # cat(as.numeric(x["start"]), " ", as.numeric(x["stop"]), "\n")
  paste0(rseq[seq(as.numeric(x["start"]), as.numeric(x["stop"]))], sep = "", collapse = "")
})
```

Use `lapply` to check the number of characters in our sequences. This is a very big virus!
```{r}
lapply(seq.fasta, nchar)
```

Get the names of the sequences from the original `rseq` object.
```{r}
names(seq.fasta) <- rseq[header.lines]
names(seq.fasta)
```

## Create a function to quickly process data
```{r}
load.genomes <- function(ids){
  rseq <- entrez_fetch("nucleotide", ids, rettype = "fasta")
  rseq <- unlist(strsplit(rseq, "\n"))
  
  header.lines <- grep(">", rseq)
  seq.pos <- cbind(start = header.lines + 1, stop = c(header.lines[-1] - 1, length(rseq)))
  seq.fasta <- apply(seq.pos, 1, function(x)
    {
    paste0(rseq[seq(as.numeric(x["start"]), as.numeric(x["stop"]))], sep = "", collapse = "")
  })
  
  names(seq.fasta) <- rseq[header.lines]
  return (seq.fasta)
}
```

Can use this function to download all genomes from NCBI.
```{r}
ncbi.seq <- load.genomes(rownames(sourceIDs))
```

```{r}
class(ncbi.seq)
length(ncbi.seq)
```

```{r}
write.csv(ncbi.seq, "C:/Users/samma/Monkeypox-genetics/data/ncbi_seq.csv")
```


# 2 Creating a blast database
```{r}
db <- ""
for (header in names(ncbi.seq)){
  cat
}
```


